### PR TITLE
Avoid creating multiple logger factory instances

### PR DESCRIPTION
**Pull request by taburet**
Fixes: https://github.com/hazelcast/hazelcast/issues/5641

Small note about the change, in the old code there was an attempt to initialize the global factory instance only if the logger type is explicitly set:

```java
loggerFactoryToUse = newLoggerFactory(loggerType);
// store the factory only when the type was set explicitly. we do not want to store default type.
if (!StringUtil.isNullOrEmpty(loggerType)) {
    loggerFactory = loggerFactoryToUse;
}
```

But seems like that was broken for about 1.5 year since `newLoggerFactory` sets `loggerFactory` anyway and nobody complained:

```java
// init static field as early as possible
if (Logger.loggerFactory == null) {
    //noinspection SynchronizationOnStaticField
    synchronized (FACTORY_LOCK) {
        if (Logger.loggerFactory == null) {
            Logger.loggerFactory = loggerFactory;
        }
    }
}
```

This PR provides the de facto behavior. If it desirable to fix the old broken behavior, I may do that.

### CODE REVIEW

**Comment by ahmetmircik** [comment]
can we have a test for the fix?

File: hazelcast/src/main/java/com/hazelcast/logging/Logger.java [hunk 1]
```diff
@@ -29,97 +35,91 @@ private Logger() {
...
-                    }
-                }
-            }
-        }
-        // loggerFactory was initialized by other thread before we acquired the lock -> loggerFactoryToUse was left to null
-        if (loggerFactoryToUse == null) {
-            loggerFactoryToUse = loggerFactory;
-        }
-        return loggerFactoryToUse.getLogger(name);
+        return getOrCreateLoggerFactory(System.getProperty("hazelcast.logging.type")).getLogger(name);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/logging/Logger.java]: Can you spread this over multiple lines. I'm not crazy about such complex expressions because it makes code harder to understand.

**Comment by taburet** [comment]
Thank you Peter for spotting such a childish mistake. Somehow I missed the fact that `loggerFactory` is initialized with `null`, not with the global shared factory instance. I will rework the solution.

**Comment by taburet** [comment]
@pveentjer what do you think about "store the factory only when the type was set explicitly" part, it was defunct for about year and a half, should I revive it?

**Comment by pveentjer** [comment]
@mdogan can you have a look at this one as well?

**Comment by blazember** [comment]
Regarding *store the factory only when the type was set explicitly*. Logger type can be configured via 1) JVM argument or 2) HZ config. I think the cited part of the original code was intended to support case 2). With the current master code in place the below snippet
```    
    public static void main(String[] args) {
        Config cfg = new Config();
        cfg.setProperty("hazelcast.logging.type", "log4j2");
        ILogger logger = Logger.getLogger(LoggerTest.class);
        HazelcastInstance instance = Hazelcast.newHazelcastInstance(cfg);
        ILogger logger2 = Logger.getLogger(LoggerTest.class);
        System.out.println(logger.getClass().getCanonicalName());
        System.out.println(logger2.getClass().getCanonicalName());
    }
```
outputs:
```
com.hazelcast.logging.StandardLoggerFactory.StandardLogger
com.hazelcast.logging.StandardLoggerFactory.StandardLogger
```

If `newLoggerFactory()` is changed to share the created `LoggerFactory` under the same conditions (the type was set explicitly), the configured logging type is respected and the output of the same code is the following:
```
com.hazelcast.logging.StandardLoggerFactory.StandardLogger
com.hazelcast.logging.Log4j2Factory.Log4j2Logger
```

In short, I think the part under question should be revived with changing `newLoggerFactory()` to share factory under the same conditions when `getLogger()` does it, so that programmatic logging configuration is respected.

**Comment by taburet** [comment]
@blazember I agree with you that the old broken behavior of `newLogger()` must be restored, but seems like the whole point of this trickery is to maximize the chances that `newLogger()` and `LoggingServiceImpl`/`ClientLoggingService` are acting on the same logger factory instance, so whatever `newLoggerFactory()` is going to return it must be stored in `Logger#loggerFactory` if it's not already set. WDYT?

**Comment by blazember** [comment]
I think it also works, as long as `getLogger()` doesn't store the created `LoggerFactory`, just reuses it if it is previously stored. Otherwise, in the above test code `Config` stores a default `StandardLogger`, if no system parameter is provided as the class definition, starts with
```
private static final ILogger LOGGER = Logger.getLogger(Config.class);
```

This also means that there is a missing test, which verifies that defining `hazelcast.logging.type` programmatically via `Config#setProperty()` is effective.

**Comment by taburet** [comment]
1. Added fixes for all observations raised during the review by @pveentjer and @blazember.

2. Added `CustomLoggerFactorySingularityTest` to verify the singularity of custom logger factories configured using `hazelcast.logging.class` as @ahmetmircik requested.

3. Added more test to `LoggerTest` to verify the expected relations between `Logger.getLogger` and `Logger.newLoggerFactory` as @blazember requested.

**Comment by taburet** [comment]
This PR will be replaced with a better one soon, closing.
